### PR TITLE
Add testset

### PR DESCRIPTION
**Pull request by Y4NG333**
Add a test set to test the model with data that is different from the training set

### CODE REVIEW

**Comment by Y4NG333** [comment]
@albert-no 교수님, 저는 main.py에 모형 저장 기능을 추가했습니다.
그리고 test.py 파일을 추가했습니다.
성능 테스트에서는 이미 훈련된 model을 도입하여 테스트할 수 있다.

File: test.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+from matplotlib import pyplot as plt
+from model import Transformer
+from utils import data_generator, DataInfo, draw_plot
+
+# Used to configure the environment
+os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"
+
+alphabet = ["a", "b", "c"]
+weight = np.array([1, 1, 1])
```
Comment (albert-no) [inline test.py]: main에서는 2, 1, 1을 썼습니다.

File: test.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+)
+criterion = nn.MSELoss()
+
+# Define model
+n_heads, d_model, n_layers, src_vocab_size, tgt_vocab_size = 8, 512, 6, len(alphabet), 3
+
+# Load model parameters
+path_model = "./model/"
+model_test = Transformer(n_heads, d_model, n_layers, src_vocab_size, tgt_vocab_size)
+model_test.load_state_dict(torch.load(path_model + str(20) + "net.pth"))
```
Comment (albert-no) [inline test.py]: path_model+str(20) .. 과 같이 쓰지말고
애초에 model.pth이름을 `fname`으로 입력을 받기 바랍니다.

File: test.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+# Generate test set
+seq, seq_int, code, code_int, code_onehot = data_generator(datainfo, 10)
```
Comment (albert-no) [inline test.py]: 여기 10은 뭔가요? 변수를 만들기 바랍니다.
이렇게 직접적으로 숫자를 사용하면 나중에 코드를 수정하기 어려워집니다.

File: test.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+seq_int = torch.LongTensor(seq_int)
+code_int = torch.LongTensor(code_int)
+outputs, enc_self_attns, dec_self_attns, dec_enc_attns = model_test(seq_int, code_int)
+real_out = [torch.argmax(x).item() for x in outputs]
+
+print("real out", np.array(real_out).reshape(-1, 12))
```
Comment (albert-no) [inline test.py]: 12도 max_len에 해당하는건가요?

File: test.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+print("code int", code_int)
+loss = criterion(outputs.float(), torch.LongTensor(code_onehot).view(-1, 3).float())
```
Comment (albert-no) [inline test.py]: 여기서 3은 뭐죠?

File: test.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from utils import data_generator, DataInfo, draw_plot
+
+# Used to configure the environment
+os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"
+
+alphabet = ["a", "b", "c"]
+weight = np.array([1, 1, 1])
+prob = weight / np.sum(weight)
+
+num_epoch = 21
```
Comment (albert-no) [inline test.py]: epoch이나 batch는 test에서 필요하지 않습니다.
불필요한 변수들은 지우기 바랍니다.

**Review by albert-no — COMMENTED** [review]
코멘트 확인하여 수정하기 바랍니다.

**Comment by Y4NG333** [comment]
교수님，안녕하세요~
네, 빨리 수정하겠습니다.

File: test.py (hunk 1 above)
Comment (Y4NG333) [reply test.py]: 숫자를 입력할 때 실수로 오타를 냈습니다.
지금 제가 수정했습니다.

File: test.py (hunk 6 above)
Comment (Y4NG333) [reply test.py]: 코드의 일부는 main.py에서 직접 복사하기 때문에 관련이 없는 변수들이 있습니다...
저는 이제 관련 없는 변수들을 모두 삭제했습니다.

File: test.py (hunk 3 above)
Comment (Y4NG333) [reply test.py]: 10은 train set의 수를 나타냅니다.
저는 10을 대체할 변수를 만들었습니다.

File: test.py (hunk 4 above)
Comment (Y4NG333) [reply test.py]: 네, 맞습니다.12는 최대 길이(max_len)입니다.

File: test.py (hunk 5 above)
Comment (Y4NG333) [reply test.py]: 출력은 1*3의 vector이기 때문에 loss를 계산할 때 출력과 같은 차원을 유지합니다.
0BF3A80F45A5FB0E58A7E65151A72802

File: test.py (hunk 2 above)
Comment (Y4NG333) [reply test.py]: 교수님, 저는 잘 모르겠습니다... fname으로 이 무엇입니까?

**Comment by Y4NG333** [comment]
@albert-no  교수님, 정말 죄송합니다. 내 컴퓨터 결함 때문에 (컴퓨터에서 자꾸 이상한 소리가 난다)이 변경 사항들은 오늘 오후에야 제출합니다.
아직 제가 잘하지 못하는 곳이 하나 있습니다.저는 model.pth이름을 무엇으로 수정해야 할지 잘 모르겠습니다.

File: test.py (hunk 2 above)
Comment (albert-no) [reply test.py]:
```
fname = "./model/20net.pth"
model_test.load_state_dict(torch.load(fname))
```
과 같이 pth파일이름을 한꺼번에 저장하는 변수를 활용하기 바랍니다

File: test.py (hunk 4 above)
Comment (albert-no) [reply test.py]: 여기도 12가 아니라 max_len을 활용하기 바랍니다.
```
print("real out", np.array(real_out).reshape(-1, max_len))
```

**Comment by albert-no** [comment]
코멘트 확인하기 바랍니다.
그리고 숫자 (3이나 12)를 직접활용하기 보다 변수를 활용하기 바랍니다.
그리고 commit을 한글로 하지말고 영어로 하기 바랍니다.

File: test.py (hunk 4 above)
Comment (Y4NG333) [reply test.py]: 네, 교수님, 수정했습니다.

File: test.py (hunk 2 above)
Comment (Y4NG333) [reply test.py]: 네, 알겠습니다!

**Comment by Y4NG333** [comment]
@albert-no 교수님, 관련 부분은 전부 수정했습니다.
아마 그때 입력 방법이 한글이었나봐요, 그래서 한글로 commit을 입력했어요. 
다음에는 전부 영어로 하겠습니다.
